src/otx/algo/utils/xai_utils.py

```diff
@@ -101,10 +103,63 @@ def convert_maps_to_dict_image(saliency_maps: np.array) -> list[dict[Any, np.arr
     return [{"map_per_image": map_per_image} for map_per_image in saliency_maps]
 
 
-def postprocess(saliency_map: np.ndarray, output_size: tuple | list = (224, 224)) -> np.ndarray:
+def postprocess(saliency_map: np.ndarray, output_size: tuple[int, int] | None) -> np.ndarray:
     """Postprocess single saliency map."""
     if saliency_map.ndim != 2:
-        raise ValueError
+        msg = "Shape mismatch."
+        raise ValueError(msg)
 
-    saliency_map = cv2.resize(saliency_map, output_size)
+    if output_size:
+        h, w = output_size
+        saliency_map = cv2.resize(saliency_map, (w, h))
     return cv2.applyColorMap(saliency_map, cv2.COLORMAP_JET)
+
+
+def dump_saliency_maps(
+    predict_result: list[OTXBatchPredEntityWithXAI | InstanceSegBatchPredEntityWithXAI | Any],
+    explain_config: ExplainConfig,
+    datamodule: EVAL_DATALOADERS | OTXDataModule,
+    output_dir: Path,
+    weight: float = 0.3,
+) -> None:
+    """Sumps saliency maps (raw and with overlay)."""
+    output_dir = output_dir / "saliency_maps"
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    for predict_result_per_batch in predict_result:
+        saliency_maps = predict_result_per_batch.saliency_maps
+        imgs_info = predict_result_per_batch.imgs_info
+        for pred_index in range(len(saliency_maps)):
+            img_id = imgs_info[pred_index].img_idx
+            img_data, image_save_name = _get_image_data_name(datamodule, img_id)
```

Comment:
I hope this will be called for every batch in the `engine.predict()`. This is because 
1. No double image decoding at dataloader and this line
```python
            img_data, image_save_name = _get_image_data_name(datamodule, img_id)
```
2. Not maintaining all the saliency maps during test workload but flushing saliency map into a disk streamly.

Reply:
This proposal make sense. Currently we just replicated the OTX1.x behavior. Optimizing it further would require noticeable engineering effort. Also we have a FF this week, so I would suggest to postpone this optimization for the future.
See my comments below for more details:

> Not maintaining all the saliency maps during test workload but flushing saliency map into a disk steamily.

We probably need to introduce a new prediction entity to handle such case for API users. Currently every data point in the prediction contains the numpy object itself. In the naive approach, it is possible to flush the maps to the disk on the per-bath basis and store a reference to them. Something more complicated can be done here as well. This is not well aligned with what we have now.

> No double image decoding at dataloader.

It is possible to register a custom callback with some logic at `on_predict_batch_end`. But the issue here is that there is only transformed batch available, although I need the original images to create an overlay. Not sure if I know an easy way to handle it fast.

Reply:
> It is possible to register a custom callback with some logic at `on_predict_batch_end`. But the issue here is that there is only transformed batch available, although I need the original images to create an overlay. Not sure if I know an easy way to handle it fast.

Why not overlay it over a transformed image? Is the saliency map computed from the transformed image I'm thinking that it is more correct that we should overlay it over the transformed image. A simple example for this is that the saliency map is computed from the transformed image with random cropping. There is no way to overlay it correctly without knowing the cropping parameters. Even if we know them, for overlaying. we just have to do the same preprocessing to make the original image same as the previously transformed image.

Anyway, I hope it to be improved in the future.